src/configLoader.js

```diff
@@ -13,34 +13,27 @@ export function configLoader(configPath) {
         return process.exit(1);
     }
 
+    const rawConfig = fs.readFileSync(configPath, 'utf8');
+    const compiledConfig = rawConfig.replace(/\$\{(.+)\}/g, (match, variableName) => {
+        const allowedNameRegex = /^[_a-zA-Z0-9]+$/;
+        if (!allowedNameRegex.test(variableName)) {
+            log.error(`Configuration variable name ${variableName} is invalid.\nAllowed characters are letters, numbers, and underscores.`);
+            process.exit(1);
+        }
+
+        if (process.env[variableName] === undefined) {
+            log.error(`Configuration value ${variableName} was not present in the environment.`);
+            process.exit(1);
+        }
+        return process.env[variableName];
+    });
+
```

Comment:
This is equivalent to the current `sed` approach but I wonder if we can use this as an opportunity to make this safer overall. With both the `sed` code and this code, if we auto-generated a password and it included special YAML characters, we could end up with a pretty hard to diagnose syntax error. It would report the syntax as being unparseable but the syntax error wouldn't appear in the file that the engineer is looking at. We could walk the entire parsed config structure and apply the regex replace to each object field. Thoughts?

Reply:
so i guess i have a couple thoughts on that:

1. could you explain a little further what you're envisioning? that seems like it'd be adding another regex step, that is, on the `process.env[variableName]`. so i think if that's true it's not quite accurate to say that we would "apply the regex replace to each object field"? but, as you say, this is an enhancement, so this isn't an objection, more a clarifying question. however...
2. if i were going to answer the question of "how should we prevent auto-generated syntax errors from creeping into our template files?" from scratch, my first instinct would be to get a library to validate the syntax for me instead of walking the whole tree. `js-yaml` doesn't appear to have an explicit `validate()` function, but it does have a `strict` option for [`safeLoad`](http://adilapapaya.com/docs/js-yaml/#safeloadstringoptions) that makes it throw an error if there's an error (not sure what it does otherwise?). but i would hope that it would handle those for us. so i can try adding `strict: true` to that call and test it. does that address the concern, if it works?
3. even if 2 doesn't work, we might still be able to avoid walking the tree. my preferred solution to this problem would be to wrap the templated variable in single quotes (and replace any occurrences of `'` in the value with `''`, as a necessary side effect, i think?) if we can. this approach does operate under the assumption that variables are strings, which seems like it would mostly hold up for Kong config values (i'm not sure how important the type of any value in-memory in Kongfig is anyway, given that they're serialized to JSON again to be sent to the Kong admin API?), but let's assume that's not true for now (and the consideration in the parenthetical). if that's the case, we could be back to walking the tree, but another approach, if we want to be extra about it, would be to enforce a typing requirement for variables. as in, we could test for the regex of `/\$[sbif]\{\}/g`, where `s`, `b`, `i`, and `f` refer to YAML scalar types. this might be an unnecessarily strict approach but it would create the same safety check without walking the whole tree. 

what do you think?

Reply:
I'd like to avoid 2 things. The first is restrictions on what the values of the environment variables can be. The second is developer confusion about yaml syntax errors. Your approach in (2) would still lead to developers being confused about the syntax error because they wouldn't see the text being parsed. Your approach in (3) gets close to addressing both. It would require the regex replacement on the environment variable value that you mention in (1). That feels complex because it requires us to account for all of the syntactic combinations allowed by yaml. Parsing the yaml and then using [mapDeep](https://deepdash.io/#mapdeep) to apply the substitution only to strings including the environment variable syntax makes the same trade-off as your (3) of limiting our environment variables to string fields but in exchange it allows us to completely ignore yaml syntax quirks.

Reply:
ahh okay i misunderstood part of what you were saying - i didn't keep all parts of your approach in my head at the same time. having `mapDeep` available addresses my core concern and makes this pretty easy.

Reply:
also, to clarify again - with my current understanding, we don't actually need the substitution i mentioned in (1)? because since we're past the YAML load, YAML syntax is irrelevant

Reply:
> also, to clarify again - with my current understanding, we don't actually need the substitution i mentioned in (1)? because since we're past the YAML load, YAML syntax is irrelevant

That's right. The changes you made for env var replacement look great to me.